### PR TITLE
K8SPXC-369 implement version service json logging

### PR DESCRIPTION
**Pull request by qJkee**
Add zap logger for grpc json requests logging

Example log
```
{"level":"info","ts":1595943775,"msg":"serving gRPC","Addr":"http://127.0.0.1:10000"}
{"level":"info","ts":1595943775,"msg":"parsed scheme: \"dns\"","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"ccResolverWrapper: sending update to cc: {[{127.0.0.1:10000  <nil> 0 <nil>}] <nil> <nil>}","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"ClientConn switching balancer to \"pick_first\"","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Channel switches to new LB policy \"pick_first\"","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Subchannel Connectivity change to CONNECTING","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Subchannel picks a new address \"127.0.0.1:10000\" to connect","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Channel Connectivity change to CONNECTING","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Subchannel Connectivity change to READY","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"Channel Connectivity change to READY","system":"grpc","grpc_log":true}
{"level":"info","ts":1595943775,"msg":"serving gRPC-Gateway and OpenAPI Documentation","gatewayAddr":"http://0.0.0.0:11000"}
{"level":"info","ts":1595943817,"msg":"server request payload logged as grpc.request.content field","system":"grpc","span.kind":"server","grpc.service":"version.VersionService","grpc.method":"Apply","grpc.request.content":{"msg":{"product":"pxc-operator","operatorVersion":"1.5.0","apply":"Latest","databaseVersion":"testDbVer","kubeVersion":"testKubeVer"}}}
{"level":"info","ts":1595943817,"msg":"finished unary call with code InvalidArgument","grpc.start_time":"2020-07-28T16:43:37+03:00","system":"grpc","span.kind":"server","grpc.service":"version.VersionService","grpc.method":"Apply","error":"rpc error: code = InvalidArgument desc = invalid current version: testDbVer","grpc.code":"InvalidArgument","grpc.time_ms":0.7229999899864197}
{"level":"info","ts":1595943841,"msg":"server request payload logged as grpc.request.content field","system":"grpc","span.kind":"server","grpc.service":"version.VersionService","grpc.method":"Apply","grpc.request.content":{"msg":{"product":"pxc-operator","operatorVersion":"1.5.0","apply":"Latest"}}}
{"level":"info","ts":1595943841,"msg":"finished unary call with code OK","grpc.start_time":"2020-07-28T16:44:01+03:00","system":"grpc","span.kind":"server","grpc.service":"version.VersionService","grpc.method":"Apply","grpc.code":"OK","grpc.time_ms":0.25699999928474426}
{"level":"info","ts":1595943841,"msg":"server response payload logged as grpc.response.content field","system":"grpc","span.kind":"server","grpc.service":"version.VersionService","grpc.method":"Apply","grpc.response.content":{"msg":{"versions":[{"product":"pxc-operator","operator":"1.5.0","matrix":{"pxc":{"8.0.19-10.1":{"imagePath":"percona/percona-xtradb-cluster:8.0.19-10.1","imageHash":"1058ae8eded735ebdf664807aad7187942fc9a1170b3fd0369574cb61206b63a","status":"recommended"}},"pmm":{"2.0.0":{"imagePath":"percona/percona-xtradb-cluster-operator:1.5.0-pmm","imageHash":"28bbb6693689a15c407c85053755334cd25d864e632ef7fed890bc85726cfb68","status":"recommended"}},"proxysql":{"2.0.12":{"imagePath":"percona/percona-xtradb-cluster-operator:1.5.0-proxysql","imageHash":"629f344e6d6b4e2cec95a461cfcba7f48759bb742182c4ebff453707aba77b6b","status":"recommended"}},"haproxy":{"2.1.7":{"imagePath":"percona/percona-xtradb-cluster-operator:1.5.0-haproxy","imageHash":"6978a9a4a49515fd8be713312221ab72729a220c180663c5bfdd5bd6d3317504","status":"recommended"}},"backup":{"8.0.13":{"imagePath":"percona/percona-xtradb-cluster-operator:1.5.0-pxc8.0-backup","imageHash":"7cddeedaa90ec9ba529340e34ee5662c9485139b2b197fc0abdbb6c261a14f62","status":"recommended"}},"operator":{"1.5.0":{"imagePath":"percona/percona-xtradb-cluster-operator:1.5.0","imageHash":"99c551f5437ed5f2fe004877c15a4665260969d0923f7c95fbf7a4c9e7305bb1","status":"recommended"}}}}]}}}
```

### CODE REVIEW

**Comment by blez** [comment]
Logs can be in file, but I believe that it also should be written to stdout

**Comment by qJkee** [comment]
> Please remove support for logging to file. Everything should be logged to stderr.
> And please add an example of actual log lines in PR description.

Requirements are changed?
Since before the main idea of this logging was logging to file for ELK stack.

Since we already have logging to console, should we merge this PR?

**Comment by AlekSi** [comment]
> Since before the main idea of this logging was logging to file for ELK stack.

I'm not aware of any plans for logging to file. Where you get that from?

All SaaS software logs to stderr. And all "modern" software in general: https://12factor.net/logs (except that we would prefer stderr instead of stdout)

**Comment by AlekSi** [comment]
> Since we already have logging to console, should we merge this PR?

Everything in the current ticket description is correct. It does not mention a file, but it mentions JSON, etc. Yes, we need that PR. Except logging to file.

And please add an example of the request logging. Currently, we don't see logs for valid requests at all. We need them.

File: main.go [hunk 1]
```diff
@@ -39,35 +44,40 @@ func getOpenAPIHandler() http.Handler {
 func main() {
 	useTLS := strings.ToLower(os.Getenv("SERVE_HTTP")) != "true"
 
-	log := grpclog.NewLoggerV2(os.Stdout, ioutil.Discard, ioutil.Discard)
-	grpclog.SetLoggerV2(log)
+	logger := initLogger()
```
Comment (blez) [inline main.go:47]: To be honest I do not like var name logger. As for me `log` is better.

File: main.go [hunk 2]
```diff
@@ -39,35 +44,40 @@ func getOpenAPIHandler() http.Handler {
...
-	log.Infof("serving gRPC on http://%s", addr)
+
+	logger.Info("serving gRPC", zap.String("Addr", "http://"+addr))
+
 	go func() {
-		log.Fatal(s.Serve(lis))
+		err := s.Serve(lis)
+		if err != nil {
+			logger.Fatal("Failed to serve grpc server", zap.Error(err))
+		}
```
Comment (blez) [inline main.go:80]: I think one line is ok here as it was.

File: main.go [hunk 3]
```diff
@@ -110,11 +120,43 @@ func main() {
 	}
 
 	if !useTLS {
-		log.Info("serving gRPC-Gateway and OpenAPI Documentation on http://", gatewayAddr)
-		log.Fatalln(gwServer.ListenAndServe())
+		logger.Info("serving gRPC-Gateway and OpenAPI Documentation", zap.String("gatewayAddr", "http://"+gatewayAddr))
+		err = gwServer.ListenAndServe()
+		if err != nil {
+			logger.Fatal("failed to serve gRPC-Gateway", zap.Error(err), zap.Bool("tls", false))
+		}
```
Comment (blez) [inline main.go:127]: Same here, one line should be ok

File: main.go [hunk 4]
```diff
@@ -110,11 +120,43 @@ func main() {
...
 	}
 
 	gwServer.TLSConfig = tlsConfig
-	log.Info("serving gRPC-Gateway and OpenAPI Documentation on https://", gatewayAddr)
-	log.Fatalln(gwServer.ListenAndServeTLS("", ""))
+	logger.Info("serving gRPC-Gateway and OpenAPI Documentation", zap.String("gatewayAddr", "https://"+gatewayAddr))
+	err = gwServer.ListenAndServeTLS("", "")
+	if err != nil {
+		logger.Fatal("failed to serve gRPC-Gateway", zap.Error(err), zap.Bool("tls", true))
+	}
```
Comment (blez) [inline main.go:135]: Same

File: main.go [hunk 5]
```diff
@@ -110,11 +120,43 @@ func main() {
 	}
 
 	if !useTLS {
-		log.Info("serving gRPC-Gateway and OpenAPI Documentation on http://", gatewayAddr)
-		log.Fatalln(gwServer.ListenAndServe())
+		logger.Info("serving gRPC-Gateway and OpenAPI Documentation", zap.String("gatewayAddr", "http://"+gatewayAddr))
+		err = gwServer.ListenAndServe()
+		if err != nil {
+			logger.Fatal("failed to serve gRPC-Gateway", zap.Error(err), zap.Bool("tls", false))
```
Comment (blez) [inline main.go:126]: Maybe I'm weird, but I do not like the fact that we need to wrap error to some `zap.Error()`
As I can see you can create use some `SugaredLogger` from zap logger which is more similar to other logger I saw. It has funcs like `Errorf`, `Infof`, etc...

But maybe I have some kind of OCD, so we can ask other peoples' thought :smile:

File: main.go (hunk 1 above)
Comment (qJkee) [reply main.go:47]: but it can be misleading, since we have a std `log` package.

File: main.go (hunk 5 above)
Comment (qJkee) [reply main.go:126]: It's not worth to to use sugared logger for it.
We use it only in few places, which is not so hard and annoying, but it give good performance

File: main.go (hunk 2 above)
Comment (qJkee) [reply main.go]: fixed fc4a972dc35840a06e1dd536ccfb1176527a749a

File: main.go (hunk 3 above)
Comment (qJkee) [reply main.go]: fixed fc4a972dc35840a06e1dd536ccfb1176527a749a

File: main.go (hunk 4 above)
Comment (qJkee) [reply main.go]: fixed fc4a972dc35840a06e1dd536ccfb1176527a749a

File: main.go (hunk 1 above)
Comment (blez) [reply main.go:47]: But we do not use it

Comment (delgod) [reply main.go:47]: I prefer `logger`, because of the hypothetical misleading issue

**Review by delgod — APPROVED** [review]

File: main.go [hunk 6]
```diff
@@ -4,20 +4,25 @@ import (
...
 	"log"
 	"mime"
 	"net"
 	"net/http"
 	"os"
 	"strings"
+	"time"
+
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+
```
Comment (AlekSi) [inline main.go:16]: Imports grouping seems a bit random there. golangci-lint would catch that.

File: main.go [hunk 7]
```diff
@@ -110,11 +119,41 @@ func main() {
...
-	log.Info("serving gRPC-Gateway and OpenAPI Documentation on https://", gatewayAddr)
-	log.Fatalln(gwServer.ListenAndServeTLS("", ""))
+	logger.Info("serving gRPC-Gateway and OpenAPI Documentation", zap.String("gatewayAddr", "https://"+gatewayAddr))
+	if err := gwServer.ListenAndServeTLS("", ""); err != nil {
+		logger.Fatal("failed to serve gRPC-Gateway", zap.Error(err), zap.Bool("tls", true))
+	}
+}
+
+func initLogger() *zap.Logger {
+	logConf := zap.NewProductionEncoderConfig()
```
Comment (AlekSi) [inline main.go:136]: That will enable JSON logging in development too, that might be too hard to read.

Feel free to adapt (but do not import) that code: https://github.com/percona-platform/saas/blob/main/pkg/logger/global.go

File: main.go (hunk 6 above)
Comment (qJkee) [reply main.go:16]: golangci-lint is not reporting anything wrong about imports
P.S goimports too

File: main.go (hunk 7 above)
Comment (qJkee) [reply main.go:136]: As one of approach
ff96bba88cd3562dec5689eac70cc0afb60816ad

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

File: main.go (hunk 6 above)
Comment (AlekSi) [reply main.go:16]: Too bad, but grouping is still random. Why is that import not in the group below, for example?

File: main.go [hunk 8]
```diff
@@ -110,11 +119,47 @@ func main() {
...
+		encoder.AppendInt64(time.Unix())
+	}
+
+	levelEnablerFunc := zap.LevelEnablerFunc(func(_ zapcore.Level) bool {
+		return true
+	})
+
+	logEncoder := zapcore.NewConsoleEncoder(logConf)
+
+	if os.Getenv("LOGGER_MODE") == "PRODUCTION" {
```
Comment (AlekSi) [inline main.go:147]: @browseman please pass that for JSON logging

**Review by AlekSi — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| api-tests | passed |

**Review by blez — APPROVED** [review]

File: main.go (hunk 8 above)
Comment (browseman) [reply main.go:145]: You want this to be set as env. variable where, dev, prod, or both?

Comment (AlekSi) [reply main.go:145]: That's for better integration with PaperTrail, so both

Comment (browseman) [reply main.go:145]: ok, will change that in the next request for deploy

**Pull request merged**